src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -250,121 +241,62 @@ public void putDataPoint(String metricName,
-                        }
-                        else {
-                            insertCountByMetric.put(metricName, 1);
-                        }
-
-                        if ((writeTime - keyInsertTimer) > 30_000) {
-                            keyInsertTimer = writeTime;
-                            keyInsertCount = 0;
-                            countsByMetricName = new ArrayList<>();
-                            countsByMetricName.addAll(insertCountByMetric.entrySet());
-                            insertCountByMetric.clear();
-                            logger.warn("RowKeys inserted: count={}", count);
-                        }
-                    }
-                }
-            }
-
-            if (writeRowKey) {
-
-                BoundStatement bs = new BoundStatement(m_psInsertRowKey);
-                bs.setBytes(0, ByteBuffer.wrap(metricName.getBytes(UTF_8)));
-
-                ByteBuffer serializedKey = DATA_POINTS_ROW_KEY_SERIALIZER.toByteBuffer(rowKey);
-
-                bs.setBytes(1, serializedKey);
-                bs.setInt(2, rowKeyTtl);
-                m_session.executeAsync(bs);
-
-                for (String split : m_indexTagList) {
-                    String v = tags.get(split);
-                    if (null == v || "".equals(v)) {
-                        continue;
-                    }
-
-                    bs = new BoundStatement(m_psInsertRowKeySplit);
-                    bs.setString(0, metricName);
-                    bs.setString(1, split);
-                    bs.setString(2, v);
-                    bs.setBytes(3, serializedKey);
-                    bs.setInt(4, rowKeyTtl);
-
-                    m_session.executeAsync(bs);
-                }
-
-                for (RowKeyListener rowKeyListener : m_rowKeyListeners) {
-                    rowKeyListener.addRowKey(metricName, rowKey, rowKeyTtl);
+            final boolean rowKeyKnown = rowKeyCache.isKnown(serializedKey);
+            if (!rowKeyKnown) {
+                storeRowKeyReverseLookups(metricName, serializedKey, rowKeyTtl, tags);
+                rowKeyCache.put(serializedKey);
+
+                int count = ++keyInsertCount;
+                if (insertCountByMetric.containsKey(metricName)) {
+                    insertCountByMetric.put(metricName, insertCountByMetric.get(metricName) + 1);
+                } else {
+                    insertCountByMetric.put(metricName, 1);
                 }
 
-                if (countsByMetricName != null) {
-                    Collections.sort(countsByMetricName, (x,y) -> y.getValue() - x.getValue());
-                    logger.warn("Keys inserted: {}", countsByMetricName.subList(0, Math.min(10, countsByMetricName.size())));
+                if ((writeTime - keyInsertTimer) > 30_000) {
+                    keyInsertTimer = writeTime;
+                    keyInsertCount = 0;
+                    countsByMetricName = new ArrayList<>(insertCountByMetric.entrySet());
```

Comment:
`insertCountByMetric` is misused here. The method is supposed to run in parallel by several threads. `insertCountByMetric` is a shared resource for them. Thus, some thread may call `insertCountByMetric.clear()` before another thread calls `insertCountByMetric.entrySet()`. That's why the lock was here initially.

What we can do here is:
1. Remove this piece of cr^W code (it serves only logging purpose anyway)
2. Extract this block outside of the function to background processing (in a parallel thread) and do logging there. Metric should be sent to that thread as messages (think of `channels` concept from Go)

Reply:
True. Completely missed it.
I'd vote for 1